core/src/main/java/org/jclouds/http/internal/BaseHttpCommandExecutorService.java

```diff
@@ -80,6 +81,64 @@ protected BaseHttpCommandExecutorService(HttpUtils utils, ContentMetadataCodec c
       this.wire = checkNotNull(wire, "wire");
       this.idempotentMethods = ImmutableSet.copyOf(idempotentMethods.split(","));
    }
+   @Override
+   public ListenableFuture<HttpResponse> invokeAsync(final HttpCommand command){
+      SettableFuture<HttpResponse> finalFuture = SettableFuture.create();
+      this.invokeAsyncInternal(command, finalFuture);
+      return finalFuture;
+   }
+
+   private void invokeAsyncInternal(final HttpCommand command, final SettableFuture<HttpResponse> finalFuture) {
+      ListenableFuture<HttpResponse> future = null;
+      try {
+         final HttpRequest request = command.getCurrentRequest();
+         final Q nativeRequest = getNativeRequest(request);
+         future = invokeAsync(nativeRequest);
+         Futures.addCallback(future, new FutureCallback<HttpResponse>() {
+            @Override
+            public void onSuccess(HttpResponse response) {
+               logResponse(request, response);
+               int statusCode = response.getStatusCode();
+               if (statusCode >= 300) {
+                  if (shouldContinue(command, response))
+                     invokeAsyncInternal(command, finalFuture);
+                  else
+                     cleanup(nativeRequest);
+                  setResultOrException(command, finalFuture, response);
+               } else {
+                  cleanup(nativeRequest);
+                  setResultOrException(command, finalFuture, response);
+               }
+            }
+
+            @Override
+            public void onFailure(Throwable t) {
+               handleInvokeAsyncFailure(t, command, finalFuture);
+            }
+         });
+      } catch (Exception e) {
+         handleInvokeAsyncFailure(e, command, finalFuture);
+
+      }
```

Comment:
I'm a bit concerned here about the duplicated logic: the retry logic and exception handling duplicate the one in the sync invoke method, with small differences to accommodate the async futures. Would it be worth refactoring the `sync` implementation to use, say, the Guava's [ImmediateFuture](https://google.github.io/guava/releases/19.0/api/docs/com/google/common/util/concurrent/Futures.html#immediateFuture(V)) so we can have just one implementation of both approaches? Would this introduce a non-negligible overhead to the `sync` methods (the common use-case) to make it not worth it?